File: python/graph/Solution207.py
# 207. 课程表
from collections import deque
from typing import List


class Solution:
    def canFinishDFS(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # pre -> course edges
        adj_list = {i: [] for i in range(numCourses)}
        for course, pre in prerequisites:
            adj_list[course].append(pre)
        # 0: not visited, 1: visiting, 2:visited
        visited = [0] * numCourses

        def dfs(node):
            if visited[node] == 2:
                return True
            if visited[node] == 1:
                return False
            if visited[node] == 0:
                visited[node] = 1
                for pre in adj_list[node]:
                    if not dfs(pre):
                        return False
                visited[node] = 2
                return True

        for i in range(numCourses):
            if not dfs(i):
                return False
        return True

    def canFinishBFS(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # pre->course edge
        adj_list = {i: [] for i in range(numCourses)}
        in_degrees = [0 for _ in range(numCourses)]
        queue = deque()
        for course, pre in prerequisites:
            adj_list[pre].append(course)
            in_degrees[course] += 1
        # all courses with indegree of 0
        for i, indegree in enumerate(in_degrees):
            if indegree == 0:
                queue.append(i)
        # bfs
        while queue:
            pre = queue.popleft()
            numCourses -= 1
            for c in adj_list[pre]:
                in_degrees[c] -= 1
                if not in_degrees[c]:
                    queue.append(c)
        return numCourses == 0

File: src/graph/Solution207.java
// 207. 课程表

package graph;

import java.util.*;

public class Solution207 {

    Map<Integer, List<Integer>> graph = new HashMap<>();
    int[] visited;

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // build graph
        for (int[] prerequisite : prerequisites) {
            int ai = prerequisite[0];
            int bi = prerequisite[1];
            List<Integer> tmp = graph.getOrDefault(ai, new ArrayList<>());
            tmp.add(bi);
            graph.put(ai, tmp);
        }
        visited = new int[numCourses];  // 0 -> unvisited, 1 -> visited, 2 -> visiting

        for (int i = 0; i < numCourses; i++) {
            if (!dfs(i)) return false;
        }

        return true;
    }

    public boolean dfs(int course) {
        if (visited[course] == 1) return true;
        if (visited[course] == 2) return false;
        List<Integer> pres = graph.get(course);
        if (pres == null) return true;
        visited[course] = 2;
        for (int pre : pres) {
            if (!dfs(pre)) {
                return false;
            }
        }
        visited[course] = 1;
        return true;
    }

    public static void main(String[] args) {
        int num = 2;
        int[][] pre = {{1, 0}};
        Solution207 sol = new Solution207();
        System.out.println(sol.canFinish(2, pre));
    }

}
